File: competitive_programming/leetcode/cpp/131_palindrome_partitioning.cpp
#include <iostream>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    vector<vector<string>> partition(string s) {
        n = s.size();
        vector<vector<string>> ans;
        vector<vector<bool>> dp(n, vector<bool>(n, true));
        constructDP(dp, s);
        vector<string> stack;
        dfs(0, s, stack, ans, dp);
        return ans;
    }

private:
    int n;

    void constructDP(vector<vector<bool>>& dp, const string& s) {
        for (int step = 1; step < n; ++step) {
            for (int i = 0; i < n; ++i) {
                int j = i + step;
                if (j >= n) {
                    break;
                }
                if (step == 1) {
                    dp[i][j] = s[i] == s[j];
                } else {
                    dp[i][j] = dp[i + 1][j - 1] && s[i] == s[j];
                }
            }
        }
    }

    void dfs(int start, string& s, vector<string>& stack, vector<vector<string>>& ans, const vector<vector<bool>>& dp) {
        if (start == n) {
            ans.push_back(stack);
            return;
        }
        for (int i = start; i < n; ++i) {
            if (dp[start][i]) {
                stack.push_back(s.substr(start, i - start + 1));
                dfs(i + 1, s, stack, ans, dp);
                stack.pop_back();
            }
        }
    }
};

void print(vector<vector<string>> vec) {
    cout << "--------------------------------\n";
    for (vector<string> line : vec) {
        for (string s : line) {
            cout << s << " ";
        }
        cout << "\n";
    }
}

int main() {
    Solution* solution = new Solution();
    vector<vector<string>> ans;

    ans = solution->partition("aab");
    print(ans);

    ans = solution->partition("a");
    print(ans);

    ans = solution->partition("aaaaaaaaaaaaaaaa");
    print(ans);
}

File: competitive_programming/leetcode/python/131_palindrome_partitioning.py
from typing import List

class Solution:
    def partition(self, s: str) -> List[List[str]]:
        dp = [[False for _ in s] for _ in s]
        n = len(s)
        for step in range(n):
            for i in range(n):
                j = i + step
                if j >= n:
                    break
                if step == 0:
                    dp[i][j] = True
                elif step == 1:
                    dp[i][j] = s[i] == s[j]
                else:
                    dp[i][j] = dp[i + 1][j - 1] and s[i] == s[j]
        ans = []
        self.dfs(0, [], ans, s, dp)
        return ans
    
    def dfs(self, start: int, stack: List[str], ans: List[List[str]], s: str, dp: List[List[bool]]) -> None:
        if start == len(s):
            ans.append(stack[:])
        for i in range(start, len(s)):
            if dp[start][i]:
                stack.append(s[start : (i + 1)])
                self.dfs(i + 1, stack, ans, s, dp)
                stack.pop()

def printAns(ans: List[List[str]]):
    print("--------------------------------")
    for line in ans:
        print(line)

solution = Solution()
ans = solution.partition("aab")
printAns(ans)


solution = Solution()
ans = solution.partition("aaaaaaaaaaaaaaaa")
print(len(ans))
